datafusion/core/src/physical_plan/windows/window_agg_exec.rs

```diff
@@ -116,14 +150,38 @@ impl ExecutionPlan for WindowAggExec {
 
     /// Get the output partitioning of this plan
     fn output_partitioning(&self) -> Partitioning {
-        // because we can have repartitioning using the partition keys
-        // this would be either 1 or more than 1 depending on the presense of
-        // repartitioning
-        self.input.output_partitioning()
+        // Although WindowAggExec does not change the output partitioning from the input, but can not return the output partitioning
+        // from the input directly, need to adjust the column index to align with the new schema.
+        let window_expr_len = self.window_expr.len();
+        let input_partitioning = self.input.output_partitioning();
+        match input_partitioning {
+            Partitioning::RoundRobinBatch(size) => Partitioning::RoundRobinBatch(size),
+            Partitioning::UnknownPartitioning(size) => {
+                Partitioning::UnknownPartitioning(size)
+            }
+            Partitioning::Hash(exprs, size) => {
+                let new_exprs = exprs
+                    .into_iter()
+                    .map(|expr| {
+                        expr.transform_down(
```

Comment:
In the same vein, it may be a good idea to use:
```rust
expr.transform_down(&|e| {
    Ok(e.as_any().downcast_ref::<Column>().map(|col| {
        Arc::new(Column::new(
            col.name(),
            window_expr_len + col.index(),
        ))
            as Arc<dyn PhysicalExpr>
    }))
})
```

Reply:
> Thank you for fixing this bug! I agree that this helps with @mustafasrepo's example, but I also agree with @metesynnada that it does not change the fact that the `output_ordering` and `output_partitioning` APIs do not work inside `create_initial_plan` (do not work in the specific sense that they may return non-final values, which can be inconsistent with other APIs like `required_output_ordering`).
> 
> So things boil down to whether we expect them to work (or not) while inside `create_initial_plan`. @mingmwang, what do you think? Should we expect them to return reliable values (e.g. consistent with `required_input_ordering()` API), or should we not expect this behavior?

Based on the current design, we can not expect them to return a fixed reliable values. Because the physical plan
tree will be adjusted and changed very dynamically, so the **real** output partitioning and output ordering is not fixed  values until pass the Enforcement rule.

For your use cases, I think if your have specific physical operators that have requirements on the ordering and partitioning, just specify the requirements through the method `required_input_ordering()` and `required_input_distribution()`. And If you have rules need to co-operator with the output partitioning and output ordering,  need to add the rules after the Enforcement rule.

Reply:
Thanks for clarifying, much appreciated. @mustafasrepo will provide a pointer to our relevant code so that you can comment.

To give you a little more context beforehand: We are implementing a pipelineable version of `WindowAggExec`, but whether we can use this more efficient version in the physical plan depends on the output ordering of the previous operator. My understanding of your suggestion is that we should do this within the context of a new rule that replaces the pipeline-breaking operator with the new pipelineable operator.

Reply:
> If you do not mind, you can also show me your code, let's take a look and figure out how to make them work and be consistent with the current design.

Sure,  we do check for the input ordering here (inside `create_initial_plan`)
> 
> https://github.com/apache/arrow-datafusion/blob/617f39e2d52915ee9cd91f9e19e44f46f567d303/datafusion/core/src/physical_plan/planner.rs#L583
>

Reply:
Sure, if I understand correctly, it is very similar to choose whether to use SortMergeJoin or HashJoin as the physical join implementation. If the input plan can produce required ordering, then prefer SortMergeJoin. Such kind of optimization is usually done in a rule.

You can also implement this rule in a more cost based matter.  For each WindowAggExec, you can either choose to
use WindowAggExec or your pipelined version, pipeline version have requirements on output ordering, the Enforcement rule will add SortExec if the input plan ordering can not satisfy. At the end, count the number of SortExecs in the tree, the tree plan which has the least number of the SortExecs is the best plan. Or you can also tolerant with 1 or 2 additional SortExec and still prefer the pipelined version.

Reply:
Great! It makes sense to go through this route -- this PR provides the necessary fix for us to implement such a rule, so our original PR is now unnecessary. We will close it and utilize this fix in a rule.